### ISSUE TITLE
Stop trying to authenticate after a number of failures

### ISSUE THREAD

**Issue by lazerlabs** [Voyz/ibeam#221]
During this night IBKR had a maintenance scheduled that broke the authentication.

However, Ibeam continued to try to authenticate and eventually my user got disabled.

After a certain amount of failed login attempts the account get disabled. Is it possible to set a "retry max N number of times than just stop"? Looking at the options I'm not sure I understand which one would achieve this.

Meanwhile, I will automatically stop the docker image for a while in the night and during the weekend.

**Comment by Voyz** [comment]
Thanks for sharing your thoughts. If it happens again, please share your logs if possible. Do you have the auto-restart on that Docker container?

Regarding this:

> However, it would be useful to have a generic login counter per day: if it goes above a certain threshold stop it. 

There is one already, it's the code I've shared with you, it stops after certain amount of failures. Unless you mean to just do a general login counter, then no, there isn't one. Would you be interested in contributing it?

**Comment by lazerlabs** [comment]
Yeah, I'll have a look.

Meanwhile it happened again. Not sure why, but the consequence at some point has been that my user has been blocked again for 24 hours. It's not a big deal, it's a secondary user used only for this, but I'd like to address the cause.

For example, beside the total number of attempts (either successful or failed), I think an exponential backoff between attempt could help too.
As you might remember I used AWS lambda to retrieve the TFA code from an SQS queue (and iPhone shortcut to send it there, via another lambda endpoint). Now, my home network today was flaky so it's possible I didn't get the code in time, but the problem is that SQS does not delete it if I don't read, so a new retry would get the old one (they are set with a TTL of 300 secs, I should probably reduce it). Anyway, a vicious circle seems to start with wrong codes, then IBKR blocks me, then ibeam cannot find the correct form because I get the message that I'm blocked.
(reference: https://github.com/lazerlabs/lambda_sqs/tree/main)

So I might need the following:
* An optional exponential backoff for any wait/retry/etc.
* A total number of allowed attempts, then stop, until we manually restart the docker. This to avoid triggering IBKR protections.

I attached the logs.
[ibeam.zip](https://github.com/user-attachments/files/17874708/ibeam.zip)

**Comment by lazerlabs** [comment]
I'll have a look how to implement both

**Comment by lazerlabs** [comment]
Ok, I investigated further, changed some variables to be as conservative as possible.

I admit I had restart set to true, but even if I set it to False, after a maintenance period, it restarts anyway.

Here's my current env variables:
IBEAM_MAX_FAILED_AUTH=1
IBEAM_RESTART_FAILED_SESSIONS=False
IBEAM_REQUEST_RETRIES=0
IBEAM_REAUTHENTICATE_WAIT=30
IBEAM_MAX_IMMEDIATE_ATTEMPTS=2
IBEAM_EXTERNAL_REQUEST_TIMEOUT=30

Now, it tries two times (although it fails both times, shouldn't it stop after 1?), and then wait for 60 seconds, and then starts over.

See the attached log.

The only way I see to stop doing this is to extend IBEAM_REAUTHENTICATE_WAIT to an exceeding amount (1 hour?), and maybe extend IBEAM_MAINTENANCE_INTERVAL to like 4 hours. 

But my question is, if I set fo False IBEAM_RESTART_FAILED_SESSION shouldn't stop after the first amount of tentatives?

**Comment by Voyz** [comment]
> Now, it tries two times (although it fails both times, shouldn't it stop after 1?),

Yes, it should stop after 1. I'd need to see the log to understand why.

> then wait for 60 seconds, and then starts over.

Yupp, that suggests the error you're encountering is not counted as a failure.

> But my question is, if I set fo False IBEAM_RESTART_FAILED_SESSION shouldn't stop after the first amount of tentatives?

IBEAM_RESTART_FAILED_SESSION controls something else, though I apologise for the nomenclature and the confusion. This variable will cause a logout and a reattempt to authenticate upon encountering a valid IBKR session that is not authenticated after a successful login.

> See the attached log.

I can't quite see any log here, could you share again?

**Comment by Bomper** [comment]
See #225, for me IBeam apparently stopped trying to log in after 5 attempts, but the process keeps going and I couldn't terminate it via Ctrl+C or SIGTERM (SIGKILL worked).

![image](https://github.com/user-attachments/assets/a196eeb5-8f85-4e8a-9e8f-3f8ccfd93f72)

**Comment by Voyz** [comment]
I'm going to close this issue due to inactivity. Thanks for your contribution and please feel free to request a reopen if you'd like to continue the discussion 👍
